### ISSUE TITLE
Source ordering mismatch in JSON/Lambdas

### ISSUE THREAD

**Issue by pauldhein** [ml4ai/delphi#200]
Sorry @pratikbhd it seems that reconciling the JSON and Lambdas for any GrFN is quickly turning into a [hydra](https://www.google.com/search?q=hydra&client=firefox-b-1-ab&source=lnms&tbm=isch&sa=X&ved=0ahUKEwi9tqad0OngAhWP2YMKHQvlD7MQ_AUIDigB#imgrc=pr5bmHiMzPdIXM:), solve one issue and several others take it's place.

Now that the variable indexing is done appropriately I have noticed that the ordering of arguments in the sources list of functions in the JSON is not in sync with the arguments as they appear in the lambda function calls. Here is an example of the problem I am seeing:

`JSON`: (function entry)
```json
{
  "name": "update_est__decision__yield_est_0",
  "type": "decision",
  "target": "yield_est",
  "reference": 11,
  "sources": [
    {
      "name": "IF_1_0",
      "type": "variable"
    },
    {
      "name": "yield_est_2",
      "type": "variable"
    },
    {
      "name": "yield_est_1",
      "type": "variable"
    }
  ]
},
```

`JSON`: (location in body of `update_est`)
```json
{
  "name": "update_est__decision__yield_est_0",
  "output": {
    "variable": "yield_est",
    "index": 3
  },
  "input": [
    {
      "variable": "IF_1",
      "index": 0
    },
    {
      "variable": "yield_est",
      "index": 2
    },
    {
      "variable": "yield_est",
      "index": 1
    }
  ]
}
```
`Lambdas`:
```python
def update_est__decision__yield_est_0(yield_est_2, yield_est_1, IF_1_0):
    return yield_est_1 if IF_1_0 else yield_est_2
```

As you can see the ordering in `input` and `sources` are are inconsistent with the ordering of the arguments in the actual lambda function. Now I could solve this by hashing the args, performing a lambda function inspection and then doing proper variable association ... but then I'd have to do that for every lambda function, and that amount of processing is non-trivial and unnecessary for us to get our GrFN representation working. Can you please update the code in `for2py` to ensure that whatever ordering of variables is used in the lambda functions is also used in **all** locations in the JSON?

**Comment by pratikbhd** [comment]
@pauldhein Yes, this is something I had noticed while I am implementing this change. While we're at this topic, the issue #185 has not been completely resolved yet, I have yet to correct the index mismatch between the pgm and lambda file as seen for the `petpt.for` file. I have posted a comment on that issue regarding an example I require to help me make the changes.

**Comment by pauldhein** [comment]
@pratikbhd per my comments on the other issue, I see how there is a problem with our current `__decision__` solution for all possible cases of `if ... {elif ...} [else ...]`, but I don't understand where the indexing issue is. Could you pull out the specific portion in the lambdas/JSON where you see the indexing issue?

**Comment by pratikbhd** [comment]
> @pratikbhd per my comments on the other issue, I see how there is a problem with our current `__decision__` solution for all possible cases of `if ... {elif ...} [else ...]`, but I don't understand where the indexing issue is. Could you pull out the specific portion in the lambdas/JSON where you see the indexing issue?

@pauldhein I have given an example of what I meant by that on issue #185. I realize now that I might have been unclear by the statement "_mismatch between the pgm and lambda file_". I apologize for the confusion caused because of this.

As for the current issue, I have made the ordering of arguments in the lambda file to align with the ordering of variables in the JSON because the JSON file is created first after which the lambda is generated, making the change of ordering of the lambda according to the JSON more convenient. Does that work for you? If the ordering of variables in the JSON has to be changed to match that of the lambda file than that can be done as well, I'll only have to add a bit more changes than I did now. 
This fix will be more robust once I can test it on the `PETPT.for` file as well (after #185 has been resolved).

**Comment by pauldhein** [comment]
@pratikbhd, as long as the variable ordering in the JSON/lambdas is in agreement then I won't be too worried about where the change occurred. It makes sense to me to make the change the way you did.

**Comment by pauldhein** [comment]
@pratikbhd after further review of the JSON that is being generated on `issue_200` I am finding another issue with the `EO` `__decision__` functions. The way we are handling this `if ... elif ...` requires us to send in the original value of `EO` to each of the decision functions so that we can set the new value to the original value in the event of an else. However the code present in the JSON includes the following:
```json
{
  "name": "petpt__decision__eo_0",
  "output": {
    "variable": "eo",
    "index": 4
  },
  "input": [
    {
      "variable": "IF_3",
      "index": 0
    },
    {
      "variable": "eo",
      "index": 1
    },
    {
      "variable": "eo",
      "index": 3
    }
  ]
},
{
  "name": "petpt__decision__eo_1",
  "output": {
    "variable": "eo",
    "index": 5
  },
  "input": [
    {
      "variable": "IF_2",
      "index": 0
    },
    {
      "variable": "eo",
      "index": 4
    },
    {
      "variable": "eo",
      "index": 2
    }
  ]
},
```
The problem here is that no `EO` input is shared between the two function calls. This is not what we want to see. If the `if` or the `elif` fails then we expect `EO` to have the value of `EO` from before the `if` checks. That value is defined in the Fortran on this line:
```fortran
EO = EEQ*1.1
```

In the JSON you have this value currently defined as `eo_1` (why is this not `eo_0`?), so the `__decision__` objects should look like this:
```json
{
  "name": "petpt__decision__eo_0",
  "output": {
    "variable": "eo",
    "index": 4
  },
  "input": [
    {
      "variable": "IF_3",
      "index": 0
    },
    {
      "variable": "eo",
      "index": 1
    },
    {
      "variable": "eo",
      "index": 2
    }
  ]
},
{
  "name": "petpt__decision__eo_1",
  "output": {
    "variable": "eo",
    "index": 5
  },
  "input": [
    {
      "variable": "IF_2",
      "index": 0
    },
    {
      "variable": "eo",
      "index": 1
    },
    {
      "variable": "eo",
      "index": 4
    }
  ]
},
```

Does this make sense? I still have some questions about the indexing scheme (it seems that some variables are growing in index faster than necessary), but we can discuss those at the next PA meeting. Lets get this change made and this PR merged.
